### PR TITLE
JIT: Regularize readbacks for parameters/OSR-locals in physical promotion

### PR DESCRIPTION
**Pull request by jakobbotsch**
Handle readbacks for parameters/OSR-locals like any other readback is
handled. Previously they were handled by creating the scratch BB and
then inserting IR after the main replacement had already been done; now,
we instead create the scratch BB eagerly and mark these as requiring a
read back at the beginning of the scratch BB, and leave normal
replacement logic up to handle it.

The main benefit is that this unification makes it easier to ensure that
future smarter handling of readbacks/writebacks (e.g. "resolution")
automatically kicks in for the common case of parameters.

Introduce another invariant, which is that we only ever mark a field as
requiring readback if it is live. Previously we would always mark them
as requiring readbacks, but would then check liveness before inserting
the actual IR to do the readback. But we don't always have the liveness
information at the point where we insert IR for readbacks after #86706.

Also expand some debug logging, and address some feedback from #86706.

### CODE REVIEW

**Comment by jakobbotsch** [comment]
jitstress restore failed with #84995... will restart it.

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib PTAL @AndyAyersMS. Few diffs as mentioned above.

[Diffs with physical promotion](https://dev.azure.com/dnceng-public/public/_build/results?buildId=298345&view=ms.vss-build-web.run-extensions-tab), [diffs without old promotion](https://dev.azure.com/dnceng-public/public/_build/results?buildId=298343&view=ms.vss-build-web.run-extensions-tab). I was mainly worried about TP of having another basic block to compute liveness for in these cases, but that doesn't seem to be costly.

File: src/coreclr/jit/promotion.cpp [hunk 1]
```diff
@@ -1617,18 +1713,18 @@ void ReplaceVisitor::WriteBackBefore(GenTree** use, unsigned lcl, unsigned offs,
 //   back before their next use.
 //
 // Parameters:
-//   lcl          - The struct local
-//   offs         - The starting offset of the range in the struct local that needs to be read back from.
-//   size         - The size of the range
+//   lcl  - Local node. Its offset is the start of the range.
+//   size - The size of the range
```
Comment (jakobbotsch) [inline src/coreclr/jit/promotion.cpp:1717]: Ugh, missing `reason` here... will address in a follow-up.

File: src/coreclr/jit/promotionliveness.cpp [hunk 2]
```diff
@@ -688,8 +688,19 @@ void PromotionLiveness::FillInLiveness(BitVec& life, BitVec volatileVars, GenTre
...
+            // Copy preexisting liveness information.
+            for (size_t i = 0; i <= agg->Replacements.size(); i++)
+            {
+                unsigned varIndex = baseIndex + (unsigned)i;
+                if (!BitVecOps::IsMember(m_bvTraits, life, varIndex))
+                {
+                    BitVecOps::AddElemD(&aggTraits, aggDeaths, (unsigned)i);
+                }
+            }
+            m_aggDeaths.Set(lcl, aggDeaths);
```
Comment (jakobbotsch) [inline src/coreclr/jit/promotionliveness.cpp:703]: We need this liveness here now to know what fields we need to mark for readback when a physically promoted struct is passed as the retbuf.

**Review by AndyAyersMS — APPROVED** [review]

**Pull request merged**